//===----------------------------------------------------------------------===//
//
// This source file is part of the Teco open source project.
//
// Copyright (c) 2022 the Teco project authors
// Licensed under Apache License v2.0
//
// See LICENSE.txt for license information
//
// SPDX-License-Identifier: Apache-2.0
//
//===----------------------------------------------------------------------===//

// THIS FILE IS AUTOMATICALLY GENERATED by TecoServiceGenerator.
// DO NOT EDIT.

extension TCAsrError {
    public struct FailedOperation: TCErrorType {
        enum Code: String {
            case errorDownFile = "FailedOperation.ErrorDownFile"
            case errorRecognize = "FailedOperation.ErrorRecognize"
            case noSuchTask = "FailedOperation.NoSuchTask"
            case serviceIsolate = "FailedOperation.ServiceIsolate"
            case userHasNoFreeAmount = "FailedOperation.UserHasNoFreeAmount"
            case userNotRegistered = "FailedOperation.UserNotRegistered"
            case other = "FailedOperation"
        }
        
        private let error: Code
        
        public let context: TCErrorContext?
        
        public var errorCode: String {
            self.error.rawValue
        }
        
        /// Initializer used by ``TCClient`` to match an error of this type.
        ///
        /// You should not use this initializer directly as there are no public initializers for ``TCErrorContext``.
        public init ?(errorCode: String, context: TCErrorContext) {
            guard let error = Code(rawValue: errorCode) else {
                return nil
            }
            self.error = error
            self.context = context
        }
        
        internal init (_ error: Code, context: TCErrorContext? = nil) {
            self.error = error
            self.context = context
        }
        
        /// 下载音频文件失败。
        public static var errorDownFile: FailedOperation {
            FailedOperation(.errorDownFile)
        }
        
        /// 识别失败。
        public static var errorRecognize: FailedOperation {
            FailedOperation(.errorRecognize)
        }
        
        /// 错误的TaskId。
        public static var noSuchTask: FailedOperation {
            FailedOperation(.noSuchTask)
        }
        
        /// 账号因为欠费停止服务，请在腾讯云账户充值。
        public static var serviceIsolate: FailedOperation {
            FailedOperation(.serviceIsolate)
        }
        
        /// 账号本月免费额度已用完。
        public static var userHasNoFreeAmount: FailedOperation {
            FailedOperation(.userHasNoFreeAmount)
        }
        
        /// 服务未开通，请在腾讯云官网语音识别控制台开通服务。
        public static var userNotRegistered: FailedOperation {
            FailedOperation(.userNotRegistered)
        }
        
        /// 操作失败。
        public static var other: FailedOperation {
            FailedOperation(.other)
        }
    }
}

extension TCAsrError.FailedOperation: Equatable {
    public static func == (lhs: TCAsrError.FailedOperation, rhs: TCAsrError.FailedOperation) -> Bool {
        lhs.error == rhs.error
    }
}

extension TCAsrError.FailedOperation: CustomStringConvertible {
    public var description: String {
        return "\(self.error.rawValue): \(message ?? "")"
    }
}

extension TCAsrError.FailedOperation {
    /// - Returns: ``TCAsrError`` that holds the same error and context.
    public func toAsrError() -> TCAsrError {
        guard let code = TCAsrError.Code(rawValue: self.error.rawValue) else {
            fatalError("Unexpected internal conversion error!\nPlease file a bug at https://github.com/teco-project/teco to help address the problem.")
        }
        return TCAsrError(code, context: self.context)
    }
}

extension TCAsrError.FailedOperation {
    /// - Returns: ``TCCommonError`` that holds the same error and context.
    public func toCommonError() -> TCCommonError? {
        if let context = self.context, let error = TCCommonError(errorCode: self.error.rawValue, context: context) {
            return error
        }
        return nil
    }
}
